src/main/scala/streamer/DataWriter.scala

```diff
@@ -0,0 +1,233 @@
+package streamer
+
+import chisel3._
+import chisel3.util._
+
+// input and output for data writer (data mover in write mode)
+class DataWriterIO(
+    dataWriterTcdmPorts: Int = DataWriterTestParameters.dataWriterTcdmPorts,
+    tcdmDataWidth: Int = DataWriterTestParameters.tcdmDataWidth,
+    unrollingDim: Int = DataWriterTestParameters.unrollingDim,
+    addrWidth: Int = DataWriterTestParameters.addrWidth,
+    fifoWidth: Int = DataWriterTestParameters.fifoWidth
+) extends Bundle {
+
+  // signals for write request address generation
+  val ptr_agu_i = Flipped(Decoupled(UInt(addrWidth.W)))
+  val unrollingStrides_csr_i = Flipped(
+    Decoupled(Vec(unrollingDim, UInt(addrWidth.W)))
+  )
+
+  // valid data from the queue
+  val data_fifo_i = Flipped(Decoupled(UInt(fifoWidth.W)))
+
+  assert(
+    fifoWidth == dataWriterTcdmPorts * tcdmDataWidth,
+    "fifoWidth should match with TCDM datawidth for now!"
+  )
+
+  // tcdm write req signals
+  val tcdm_req_addr = Output(Vec(dataWriterTcdmPorts, UInt(addrWidth.W)))
+  val write_tcmd_valid_o = Output(Vec(dataWriterTcdmPorts, Bool()))
+  val tcdm_req_data = Output(Vec(dataWriterTcdmPorts, UInt(tcdmDataWidth.W)))
+
+  val tcdm_ready_i = Input(Vec(dataWriterTcdmPorts, Bool()))
+
+  // from temporal address generation unit to indicate if the transaction is done
+  val data_move_done = Input(Bool())
+
+}
+
+// data writer, for sending write request to TCDM, getting valid data from the queue
+// data consumer from the accelerator X aspect
+class DataWriter(
+    dataWriterTcdmPorts: Int = DataWriterTestParameters.dataWriterTcdmPorts,
+    tcdmDataWidth: Int = DataWriterTestParameters.tcdmDataWidth,
+    unrollingDim: Int = DataWriterTestParameters.unrollingDim,
+    unrollingFactor: Seq[Int] = DataWriterTestParameters.unrollingFactor,
+    addrWidth: Int = DataWriterTestParameters.addrWidth,
+    fifoWidth: Int = DataWriterTestParameters.fifoWidth,
+    elementWidth: Int = DataWriterTestParameters.elementWidth
```

Comment:
Same here, so much code duplication!
```scala
class DataReader(
    dataReaderTcdmPorts: Int = DataReaderTestParameters.dataReaderTcdmPorts,
    tcdmDataWidth: Int = DataReaderTestParameters.tcdmDataWidth,
    unrollingDim: Int = DataReaderTestParameters.unrollingDim,
    unrollingFactor: Seq[Int] = DataReaderTestParameters.unrollingFactor,
    addrWidth: Int = DataReaderTestParameters.addrWidth,
    fifoWidth: Int = DataReaderTestParameters.fifoWidth,
    elementWidth: Int = DataReaderTestParameters.elementWidth
```

Reply:
We will refactor this later because this will affect all the files in other branches.